[CI] InternalCartesianCentroidTests testEqualsAndHashcode failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/2douff6uor2wy/tests/:x-pack:plugin:spatial:test/org.elasticsearch.xpack.spatial.search.aggregations.metrics.InternalCartesianCentroidTests/testEqualsAndHashcode

**Reproduction line:**
`./gradlew ':x-pack:plugin:spatial:test' --tests "org.elasticsearch.xpack.spatial.search.aggregations.metrics.InternalCartesianCentroidTests.testEqualsAndHashcode" -Dtests.seed=8D68DA4C3F1BC51F -Dtests.locale=tr -Dtests.timezone=Etc/GMT+4 -Druntime.java=17`

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.spatial.search.aggregations.metrics.InternalCartesianCentroidTests&tests.test=testEqualsAndHashcode

**Failure excerpt:**
```
java.lang.AssertionError: InternalCartesianCentroid mutation should not be equal to original
Expected: not <InternalCentroid{centroid=3.402823466385285E38, 0.0, count=139}>
     but: was <InternalCentroid{centroid=3.402823466385285E38, 0.0, count=139}>

  at __randomizedtesting.SeedInfo.seed([8D68DA4C3F1BC51F:FC67A281F0FC8C30]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.elasticsearch.test.EqualsHashCodeTestUtils.checkEqualsAndHashCode(EqualsHashCodeTestUtils.java:73)
  at org.elasticsearch.test.AbstractWireTestCase.testEqualsAndHashcode(AbstractWireTestCase.java:58)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:833)

```

--- a/test/framework/src/main/java/org/elasticsearch/test/AbstractWireTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/AbstractWireTestCase.java
@@ -53,7 +53,7 @@ public abstract class AbstractWireTestCase<T> extends ESTestCase {
      * Tests that the equals and hashcode methods are consistent and copied
      * versions of the instance are equal.
      */
-    public void testEqualsAndHashcode() {
+    public final void testEqualsAndHashcode() {
         for (int runs = 0; runs < NUMBER_OF_TEST_RUNS; runs++) {
             EqualsHashCodeTestUtils.checkEqualsAndHashCode(createTestInstance(), this::copyInstance, this::mutateInstance);
         }

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/InternalCartesianCentroidTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/InternalCartesianCentroidTests.java
@@ -94,13 +94,6 @@ public class InternalCartesianCentroidTests extends InternalAggregationTestCase<
         assertEquals(sampled.count(), samplingContext.scaleUp(reduced.count()), 0);
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/90474")
-    @Override
-    public void testEqualsAndHashcode() {
-        // TODO: When this issue is fixed, remove this method and set the parent method back to final
-        super.testEqualsAndHashcode();
-    }
-
     public void testReduceMaxCount() {
         InternalCartesianCentroid maxValueCentroid = new InternalCartesianCentroid(
             "agg",

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/InternalCartesianCentroidTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/InternalCartesianCentroidTests.java
@@ -27,6 +27,8 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
 
 import static org.hamcrest.Matchers.closeTo;
 import static org.hamcrest.Matchers.equalTo;
@@ -54,8 +56,8 @@ public class InternalCartesianCentroidTests extends InternalAggregationTestCase<
     protected InternalCartesianCentroid createTestInstance(String name, Map<String, Object> metadata) {
         Point point = ShapeTestUtils.randomPoint(false);
         CartesianPoint centroid = new CartesianPoint(point.getX(), point.getY());
-        // TODO: investigate need for encode/decode to prevent hashcode test failures
-
+        // Unlike InternalGeoCentroid, we do not need to encode/decode to handle hashcode test failures,
+        // but we do need to treat zero values with care. See the mutate function below for details on that
         long count = randomIntBetween(0, 1000);
         if (count == 0) {
             centroid = null;
@@ -145,9 +147,9 @@ public class InternalCartesianCentroidTests extends InternalAggregationTestCase<
                 } else {
                     CartesianPoint newCentroid = new CartesianPoint(centroid);
                     if (randomBoolean()) {
-                        newCentroid.resetY(centroid.getY() / 2.0);
+                        mutateCoordinate(centroid::getY, newCentroid::resetY);
                     } else {
-                        newCentroid.resetX(centroid.getX() / 2.0);
+                        mutateCoordinate(centroid::getX, newCentroid::resetX);
                     }
                     centroid = newCentroid;
                 }
@@ -163,5 +165,18 @@ public class InternalCartesianCentroidTests extends InternalAggregationTestCase<
             default -> throw new AssertionError("Illegal randomisation branch");
         }
         return new InternalCartesianCentroid(name, centroid, count, metadata);
+    }
+
+    /**
+     * The previous mutation of dividing by 2.0 left zero values unchanged, leading to lack of mutation.
+     * Now we act differently on small values to ensure that mutation actually occurs.
+     */
+    private void mutateCoordinate(Supplier<Double> getter, Consumer<Double> setter) {
+        double coordinate = getter.get();
+        if (Math.abs(coordinate) < 1e-6) {
+            setter.accept(coordinate + 1.0);
+        } else {
+            setter.accept(coordinate / 2.0);
+        }
     }
 }

